File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/job/Context.java
/*
 *
 *  Copyright (c) 2007, 2008, Cemagref
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation; either version 3 of
 *  the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */
package org.openmole.core.implementation.job;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.core.implementation.data.Variable;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;


import org.openmole.core.model.job.IContext;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.commons.tools.pattern.IVisitable;
import org.openmole.commons.tools.pattern.IVisitor;
import org.openmole.core.model.data.IVariable;
import org.openmole.core.model.data.IPrototype;

public class Context implements IVisitable<IVariable>, IContext {

    final Map<String, IVariable> variables;
 
    public Context() {
        variables = new TreeMap<String, IVariable>();
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#setVariables(java.util.ArrayList)
     */
    @Override
    public void putVariables(Iterable<? extends IVariable> vars) {
        for (IVariable<?> variable : vars) {
            putVariable(variable);
        }
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#getVariables()
     */
    @Override
    public Map<String, IVariable> getVariables() {
        return variables;
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#getVariable(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public <T> IVariable<? extends T> getVariable(IPrototype<T> proto) {
        return (IVariable<T>) getVariables().get(proto.getName());
    }

    /* @Override*/
    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#putVariable(org.openmole.core.data.Variable)
     */
    @Override
    public void putVariable(IVariable<?> variable) {
        getVariables().put(variable.getPrototype().getName(), variable);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#putVariable(java.lang.String, java.lang.Object)
     */
    @Override
    public void putVariable(String name, Object value) {
        putVariable(new Variable<Object>(name, value));
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#putVariable(java.lang.String, java.lang.Class, java.lang.Object)
     */
    @Override
    public <T> void putVariable(String name, Class<? super T> type, T value) {
        putVariable(new Variable<Object>(name, type, value));
    }
    

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#putVariable(org.openmole.core.data.structure.Prototype, java.lang.Object)
     */
    @Override
    public <T> void putVariable(IPrototype<? super T> proto, T value) {
        putVariable(new Variable<T>(proto, value));
    }

    //	@Override
	/* (non-Javadoc)
     * @see org.openmole.core.task.IContext#getLocalValue(java.lang.String)
     */
    @Override
    public <IT> IT getValue(String name) {
        synchronized (getVariables()) {
            Variable<IT> elt = getVariable(name);

            if (elt != null) {
                return elt.getValue();
            } else {
                return null;
            }
        }
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#getLocalValue(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public <IT> IT getValue(IPrototype<IT> proto) {
        return this.<IT>getValue(proto.getName());
    }

    //	@Override
	/* (non-Javadoc)
     * @see org.openmole.core.task.IContext#getLocalVariable(java.lang.String)
     */
    @Override
    public <IT> Variable<IT> getVariable(String name) {
        return (Variable<IT>) getVariables().get(name);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#removeVariable(java.lang.String)
     */
    @Override
    public void removeVariable(String name) {
        getVariables().remove(name);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#exist(java.lang.String)
     */
    @Override
    public boolean containsVariableWithName(String name) {
        return getVariables().containsKey(name);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#exist(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public boolean containsVariableWithName(IPrototype<?> proto) {
        return getVariables().containsKey(proto.getName());
    }


    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#exist(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public boolean contains(IPrototype<?> proto) {
        Variable<?> v = getVariable(proto.getName());
        if (v == null) {
            return false;
        }
        return proto.getType().isAssignableFrom(v.getType());
    }
    
    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#setValue(java.lang.String, java.lang.Object)
     */
    @Override
    public void setValue(String name, Object value) {
        synchronized (getVariables()) {
            Variable var = getVariable(name);

            if (var != null) {
                var.setValue(value);
            } else {
                putVariable(name, value);
            }
        }
    }


    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#setValue(org.openmole.core.data.structure.Prototype, java.lang.Object)
     */
    @Override
    public <T> void setValue(IPrototype<? super T> proto, T value) {
        synchronized (getVariables()) {
            IVariable var = getVariable(proto);

            if (var != null) {
                var.setValue(value);
            } else {
                putVariable(proto, value);
            }

        }
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#clean()
     */
    @Override
    public void clean() {
        synchronized (getVariables()) {
            variables.clear();
        }
    }

    /* (non-Javadoc)
     * @see org.openmole.core.task.IContext#visit(org.openmole.commons.tools.IVisitor)
     */
    @Override
    public void visit(IVisitor<IVariable> visitor) throws InternalProcessingError, UserBadDataError {
        for (IVariable<?> v : getVariables().values()) {
            visitor.action(v);
        }
    }


    /*@Override
    public Collection<IChildContext> getChlids() {
    return childs.values();
    }*/
    @Override
    public Iterator<IVariable> iterator() {
        return variables.values().iterator();
    }

}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/data/Context.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.data

import org.openmole.core.model.data.{IContext,IVariable,IPrototype, IData}
import scala.collection.immutable.TreeMap

object Context {
  
  val empty = new Context(TreeMap.empty)
  
  def apply(variables: IVariable[_]*): Context = apply(variables.toIterable)
  def apply(variables: Iterable[IVariable[_]]): Context = new Context(TreeMap.empty[String, IVariable[_]] ++ variables.map{v => v.prototype.name -> v})
  def apply(context: IContext, variables: java.lang.Iterable[IVariable[_]]) = {
    import collection.JavaConversions._
    context ++ variables
  }
  
  implicit def decorateVariableIterable(variables: Iterable[IVariable[_]]) = new {
    def toContext = Context(variables)
  }
  
}

class Context(val variables: TreeMap[String, IVariable[_]]) extends Map[String, IVariable[_]] with IContext {

  def this() = this(TreeMap.empty)

  override def +[B1 >: IVariable[_]](kv: (String, B1)) = variables + kv
  //override def +(name: String, variable: IVariable[_]) = new Context(variables + (name -> variable))
  override def -(name: String) = new Context(variables - name)
  
  //override def +(name: String, value: Object) = new Context(variables + (name -> new Variable[Object](name, value)))
    
  //override def +[T](name: String, t: Class[T], value: T) = new Context(variables + (name -> new Variable[T](name, t, value)))
 
  override def +[T] (p: IPrototype[T], value: T) = new Context(variables + (p.name -> new Variable[T](p, value)))
  
  override def +[T] (v: IVariable[T]) = new Context(variables + (v.prototype.name -> v))
  
  override def +(ctx: IContext) = new Context(variables ++ ctx)
  
  override def ++(vs: Traversable[IVariable[_]]): IContext = new Context(variables ++ (vs.map{v => v.prototype.name -> v}))
  
  override def --(names: Traversable[String]): IContext = new Context(variables -- names)
  
  override def value[T](name: String): Option[T] = variables.get(name) match {
    case None => None
    case Some(v) => Some(v.asInstanceOf[IVariable[T]].value)
  }

  override def value[T](proto: IPrototype[T]) = value[T](proto.name)

  override def variable[T](proto: IPrototype[T]): Option[IVariable[T]] = {
    variables.get(proto.name) match {
      case None => None
      case Some(v) => Some(v.asInstanceOf[IVariable[T]])
    }
  }
    
  override def variable[T](name: String): Option[IVariable[T]] = variables.get(name).asInstanceOf[Option[IVariable[T]]]
 
  override def contains(p: IPrototype[_]): Boolean = {
    variable(p.name) match {
      case None => false
      case Some(v) => p.isAssignableFrom(v.prototype)
    }
  }

  override def empty = Context.empty
 
  override def get(key: String) = variables.get(key)
  
  override def iterator = variables.iterator
  
  override def toString = "{" + (if(variables.values.isEmpty) "" else variables.values.mkString(", ")) + "}"
}
